### DISCUSSION TITLE
List of objects with required values

### DISCUSSION THREAD

**Discussion by Siecje** [json-schema-org/community#191] [Q&A] [answered]
How do you define required values in a list?

I have a list of objects with a property which is an enum with 10 options. But the list must contain at least 4 items with specific field values.

For example if the list is for contact information, the **field** property of each list object can be `phone_number`, or `cell_number`, or `address`, or `first_name`, `last_name`, `parent1_name`, etc.

The list must have a telephone number either `phone_number` or `cell_number`, but not both and then it must contain `address`, `first_name` and `last_name` and other fields are optional.

```
contact_information: [
  {
    field: "phone_number",
    value: "",
  },
]
```

**Comment by karenetheridge** [comment, marked as answer]
```json
"type": "array",
"items": { // this schema applies to all items
  "type": "object",
  "required": [ "address", "first_name", "last_name" ],
  "properties": {
    "address": { "type": "string", ... },
    .. all your other property definitions here...
  },
  "oneOf": [
    { "required": [ "phone_ number" ] },
    { "required": [ "cell_ number" ] }
  ]
}
```

See https://json-schema.org/understanding-json-schema/reference/combining.html and https://json-schema.org/understanding-json-schema/reference/array.html

**Reply by Siecje** [reply]
I believe this works for making one of two fields required but how can you specify certain values of an list of enums are required?

**Reply by Siecje** [reply]
The problem is they are not different fields.

It is a list of objects with `field` having different values.

```
contact_information: [
  {
    field: "phone_number",
    value: "",
  },
{
    field: "cell_number",
    value: "",
  },
{
    field: "address",
    value: "",
  },
{
    field: "first_name",
    value: "",
  },
]
```


```
"contact_information": {
            "type": "array",
            "items": {
                "type": "object",
                "properties": {
                    "field": {
                        "enum": [
                            "first_name",
                            "last_name",
                            "address",
                            "phone_number",
                            "cell_number",
                        ]
                    },
```

**Reply by karenetheridge** [reply]
oh I see, they are not property names, but property values.

Then you would do this:

```
"type": "array",
"items": {
  "type": "object",
  "required": [ "field", "value" ],
  "properties": {
    "field": { "enum": [ "first_name", "last_name", ... ] },
    "value": { "type": "string" }
  },
},
"allOf": [
  { "contains": { "properties": { "field": { "const": "first_name" } } } },
  { "contains": { "properties": { "field": { "const": "last_name" } } } },
  { "contains": { "properties": { "field": { "const": "address_name" } } } }
],
"oneOf": [
    { "contains": { "properties": { "field": { "const": "phone_number" } } } },
    { "contains": { "properties": { "field": { "const": "cell_number" } } } }
]
```

You should be able to find the rest of what you need at the documentation site I listed above. Try experimenting and see what happens!
